src/geant4/DagSolid.cc

```diff
@@ -311,61 +345,92 @@ G4double DagSolid::DistanceToIn(const G4ThreeVector& p) const {
+
 G4double DagSolid::DistanceToOut(const G4ThreeVector& p, const G4ThreeVector& v,
                                  const G4bool calcNorm, G4bool* validNorm,
                                  G4ThreeVector* n) const {
-  G4double minDist = kInfinity;
   double position[3] = {p.x() / cm, p.y() / cm,
                         p.z() / cm};  // convert position to cm
-  double dir[3] = {v.x(), v.y(), v.z()};
+  G4ThreeVector vec = v.unit();
+  double dir[3] = {vec.x(), vec.y(), vec.z()};
 
-  EntityHandle next_surf;
-  double next_dist;
-  DagMC::RayHistory history;
+  EntityHandle surface;
+  G4double distance;
+  moab::DagMC::RayHistory history;
 
-  fdagmc->ray_fire(fvolEntity, position, dir, next_surf, next_dist, &history, 0,
+  // fire a ray
+  fdagmc->ray_fire(fvolEntity, position, dir, surface, distance, &history, 0,
                    1);
-  history.reset();
-  next_dist *= cm;  // convert back to mm
-
-  // no more surfaces
-  if (next_surf == 0) return kInfinity;
 
+  // if we are asked to calculate the normal
   if (calcNorm) {
-    *n = SurfaceNormal(p + minDist * v);
-    *validNorm = false;
+    *validNorm = true;
+    // get the direction vector;
+    double normal[3];
+    double hit[3];
+    hit[0] = position[0] + (distance * dir[0]);
+    hit[1] = position[1] + (distance * dir[1]);
+    hit[2] = position[2] + (distance * dir[2]);
+
+    fdagmc->get_angle(surface, hit, normal, &history);
+    *n = G4ThreeVector(normal[0], normal[1], normal[2]);
+    G4int sense;
+    fdagmc->surface_sense(fvolEntity, surface, sense);
+    if (sense == -1) *n = -(*n);
   }
 
-  if (next_dist < minDist) minDist = next_dist;
+  // if hit is too close
+  if (distance * cm <= kCarToleranceHalf) {
+    distance = 0.;
+  }
 
-  // particle considered to be on surface
-  if (minDist > 0.0 && minDist <= 0.5 * kCarTolerance) {
-    return 0.0;
-  } else if (minDist < kInfinity) {
-    if (calcNorm)  // if calc norm true,s should set validNorm true
-      *validNorm = true;
-    return minDist;
-  } else {
-    return 0.0;  // was kinfinity
+#ifdef DAGDEBUG
+  G4cout << "<<<<<" << G4endl;
+  G4cout << "DistanceToOut(p,v) " << G4endl;
+  G4cout << "Name: " << Myname << G4endl;
+  G4cout << "point: " << p << G4endl;
+  G4cout << "direction: " << v << G4endl;
+  G4cout << "distance: " << distance * cm << G4endl;
+  G4cout << "calcNorm: " << calcNorm << G4endl;
+  if (calcNorm) {
+    G4cout << "validNorm: " << *validNorm << G4endl;
+    G4cout << "normal: " << n->x() << " " << n->y() << " " << n->z() << G4endl;
   }
+  G4cout << ">>>>>" << G4endl;
+#endif
+
+  // return the hit distance
+  return distance * cm;
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 //
 // G4double DistanceToOut(const G4ThreeVector& p)
 //
 // Calculate distance to nearest surface of shape from an inside point.
-
+// The distance can be an underestimate.
 G4double DagSolid::DistanceToOut(const G4ThreeVector& p) const {
```

Comment:
Other than the name of the function in the debug string, this is identical to `DistanceToIn()` - should we have one function that does this and wrap it in `DistanceToIn()` and `DistanceToOut()` to ensure consistency while preserving the API and semantics?